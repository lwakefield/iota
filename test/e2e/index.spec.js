/* eslint-env jest */
import Vdoom from '../../src'

describe('simple todo app', () => {
  document.body.innerHTML = `
    <div id="app">
      <new-todo></new-todo>
      <todo i-for="t of todos" :id="$index"></todo>
    </div>

    <template id="new-todo">
      <form>
        <input type="text" value="\${todo}" />
        <button @click="add" name="new-todo">Add</button>
      </form>
    </template>

    <template id="todo">
      <div>
        \${id} - \${todo}
        <button @click="remove" name="remove-todo"></button>
      </div>
    </template>
  `

  const todos = []

  class NewTodo extends Vdoom.Component {
    static $template = document.querySelector('#new-todo')
    $data = {todo: '', todos}

    add () {
      this.todos.push(this.todo)
      this.todo = ''
    }
  }
  Vdoom.registerComponent('new-todo', NewTodo)

  class Todo extends Vdoom.Component {
    static $template = document.querySelector('#todo')
    $data = {todo: todos[this.id], todos}

    remove () {
      this.todos.splice(this.id, 1)
    }
  }
  Vdoom.registerComponent('todo', Todo)

  class App extends Vdoom.Component {
    static $template = document.querySelector('#app')
    $data = {todos}
  }
  const app = new App
  app.mount(document.querySelector('#app'))

  it('functions correctly', () => {
    expect(document.body.outerHTML).toMatchSnapshot()

    document.querySelector('input').value = 'foo'
    document.querySelector('input').dispatchEvent(new Event('input'))
    document.querySelector('button[name="new-todo"]')
      .dispatchEvent(new Event('click'))
    expect(document.body.outerHTML).toMatchSnapshot()

    document.querySelector('input').value = 'bar'
    document.querySelector('input').dispatchEvent(new Event('input'))
    document.querySelector('button[name="new-todo"]')
      .dispatchEvent(new Event('click'))
    expect(document.body.outerHTML).toMatchSnapshot()

    document.querySelector('button[name="remove-todo"]')
      .dispatchEvent(new Event('click'))
    expect(document.body.outerHTML).toMatchSnapshot()
  })
})
