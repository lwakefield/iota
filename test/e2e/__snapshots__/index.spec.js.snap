exports[`simple todo app functions correctly 1`] = `
"<body>
    <div id=\"app\">
      <form>
        <input type=\"text\" value=\"\">
        <button name=\"new-todo\">Add</button>
      </form>
      
    </div>

    <template id=\"new-todo\">
      <form>
        <input type=\"text\" value=\"\${todo}\">
        <button @click=\"add\" name=\"new-todo\">Add</button>
      </form>
    </template>

    <template id=\"todo\">
      <div>
        \${id} - \${todo}
        <button @click=\"remove\" name=\"remove-todo\"></button>
      </div>
    </template>
  </body>"
`;

exports[`simple todo app functions correctly 2`] = `
"<body>
    <div id=\"app\">
      <form>
        <input type=\"text\" value=\"\">
        <button name=\"new-todo\">Add</button>
      </form>
      <div>
        0 - foo
        <button name=\"remove-todo\"></button>
      </div>
    </div>

    <template id=\"new-todo\">
      <form>
        <input type=\"text\" value=\"\${todo}\">
        <button @click=\"add\" name=\"new-todo\">Add</button>
      </form>
    </template>

    <template id=\"todo\">
      <div>
        \${id} - \${todo}
        <button @click=\"remove\" name=\"remove-todo\"></button>
      </div>
    </template>
  </body>"
`;

exports[`simple todo app functions correctly 3`] = `
"<body>
    <div id=\"app\">
      <form>
        <input type=\"text\" value=\"\">
        <button name=\"new-todo\">Add</button>
      </form>
      <div>
        0 - foo
        <button name=\"remove-todo\"></button>
      </div><div>
        1 - bar
        <button name=\"remove-todo\"></button>
      </div>
    </div>

    <template id=\"new-todo\">
      <form>
        <input type=\"text\" value=\"\${todo}\">
        <button @click=\"add\" name=\"new-todo\">Add</button>
      </form>
    </template>

    <template id=\"todo\">
      <div>
        \${id} - \${todo}
        <button @click=\"remove\" name=\"remove-todo\"></button>
      </div>
    </template>
  </body>"
`;

exports[`simple todo app functions correctly 4`] = `
"<body>
    <div id="app">
      <form>
        <input type="text" value="">
        <button name="new-todo">Add</button>
      </form>
      <div>
        0 - foo
        <button name="remove-todo"></button>
      </div>
    </div>

    <template id="new-todo">
      <form>
        <input type="text" value="\${todo}">
        <button @click="add" name="new-todo">Add</button>
      </form>
    </template>

    <template id="todo">
      <div>
        \${id} - \${todo}
        <button @click="remove" name="remove-todo"></button>
      </div>
    </template>
  </body>"
`;
